File: stream-chat-android-docs/src/main/java/io/getstream/chat/docs/java/client/docusaurus/Debugging.java
package io.getstream.chat.docs.java.client.docusaurus;

import android.content.Context;

import androidx.annotation.NonNull;

import io.getstream.chat.android.client.ChatClient;
import io.getstream.chat.android.client.debugger.ChatClientDebugger;
import io.getstream.chat.android.client.debugger.SendMessageDebugger;
import io.getstream.chat.android.models.Message;
import io.getstream.result.Error;
import io.getstream.result.Result;

/**
 * @see <a href="https://getstream.io/chat/docs/sdk/android/basics/debugging/">Debugging</a>
 */
public class Debugging {

    public void addClientDebugger(@NonNull Context context) {
        final ChatClient chatClient = new ChatClient.Builder("apiKey", context)
                .clientDebugger(new ChatClientDebugger() {
                    @Override
                    public void onNonFatalErrorOccurred(@NonNull String tag, @NonNull String src, @NonNull String desc, @NonNull Error error) {
                        // TODO: Implement your custom logic here
                    }

                    @NonNull
                    @Override
                    public SendMessageDebugger debugSendMessage(
                            @NonNull String channelType,
                            @NonNull String channelId,
                            @NonNull Message message,
                            boolean isRetrying
                    ) {
                        return Debugging.this.debugSendMessage(
                                channelType,
                                channelId,
                                message,
                                isRetrying
                        );
                    }
                })
                .build();
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/sdk/android/basics/debugging/#debug-message-sending">Debug Message Sending</a>
     */
    private SendMessageDebugger debugSendMessage(
            @NonNull String channelType,
            @NonNull String channelId,
            @NonNull Message message,
            boolean isRetrying
    ) {
        return new SendMessageDebugger() {

            @Override
            public void onStart(@NonNull Message message) {
                // handle onStart
            }

            @Override
            public void onInterceptionStart(@NonNull Message message) {
                // handle onInterceptionStart
            }

            @Override
            public void onInterceptionUpdate(@NonNull Message message) {
                // handle onInterceptionUpdate
            }

            @Override
            public void onInterceptionStop(@NonNull Result<Message> result, @NonNull Message message) {
                // handle onInterceptionStop
            }


            @Override
            public void onSendStart(@NonNull Message message) {
                // handle onSendStart
            }

            @Override
            public void onSendStop(@NonNull Result<Message> result, @NonNull Message message) {
                // handle onSendStop
            }


            @Override
            public void onStop(@NonNull Result<Message> result, @NonNull Message message) {
                // handle onStop
            }

        };
    }

}

File: stream-chat-android-docs/src/main/kotlin/io/getstream/chat/docs/kotlin/client/docusaurus/Debugging.kt
package io.getstream.chat.docs.kotlin.client.docusaurus

import android.content.Context
import io.getstream.chat.android.client.ChatClient
import io.getstream.chat.android.client.debugger.ChatClientDebugger
import io.getstream.chat.android.client.debugger.SendMessageDebugger
import io.getstream.chat.android.models.Message
import io.getstream.result.Error

import io.getstream.result.Result

/**
 * @see <a href="https://getstream.io/chat/docs/sdk/android/basics/debugging/">Debugging</a>
 */
class Debugging {

    fun addClientDebugger(context: Context) {
        val client = ChatClient.Builder("apiKey", context)
            .clientDebugger(object : ChatClientDebugger {
                override fun onNonFatalErrorOccurred(tag: String, src: String, desc: String, error: Error) {
                    // TODO: Implement your custom logic here
                }

                override fun debugSendMessage(
                    channelType: String,
                    channelId: String,
                    message: Message,
                    isRetrying: Boolean,
                ): SendMessageDebugger = this@Debugging.debugSendMessage(
                    channelType,
                    channelId,
                    message,
                    isRetrying
                )
            })
            .build()
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/sdk/android/basics/debugging/#debug-message-sending">Debug Message Sending</a>
     */
    private fun debugSendMessage(
        channelType: String,
        channelId: String,
        message: Message,
        isRetrying: Boolean = false,
    ): SendMessageDebugger = object : SendMessageDebugger {
        override fun onStart(message: Message) {
            // handle onStart
        }

        override fun onInterceptionStart(message: Message) {
            // handle onInterceptionStart
        }

        override fun onInterceptionUpdate(message: Message) {
            // handle onInterceptionUpdate
        }

        override fun onInterceptionStop(result: Result<Message>, message: Message) {
            // handle onInterceptionStop
        }

        override fun onSendStart(message: Message) {
            // handle onSendStart
        }

        override fun onSendStop(result: Result<Message>, message: Message) {
            // handle onSendStop
        }

        override fun onStop(result: Result<Message>, message: Message) {
            // handle onStop
        }
    }
}